lib/srv/alpnproxy/local_proxy.go

```diff
@@ -158,7 +158,16 @@ func (l *LocalProxy) handleDownstreamConnection(ctx context.Context, downstreamC
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	defer upstreamConn.Close()
+	defer tlsConn.Close()
+	if err := tlsConn.Handshake(); err != nil {
```

Comment:
is `Handshake` required?

Reply:
This call guarantees that we have the `NegotiatedProtocol`  from the connection state (used to define if the connection needs to be wrapped or not) before performing any Read/Write.

Reply:
ok. i could be wrong but i thought `Dial` already handshakes

Reply:
I can do a quick test to guarantee it is necessary, but as the docs state:

> Handshake runs the client or server handshake protocol if it has not yet been run.
Most uses of this package need not call Handshake explicitly: the first Read or Write will call it automatically.

Reply:
Yeah, I confirmed (with a test and looking at the code) that Dial performs the `Handshake`. I'll remove it. Thanks for pointing it out.